libs/ml/include/ml/ops/ops.hpp

```diff
@@ -149,6 +150,57 @@ class Ops
     // Empty deafult implementation for non-trainable Ops.
   }
 
+  /**
+   * @brief ChargeForward
+   * @return estimated charge amount, necessary for performing a forward pass on data of given
+   * shapes.
+   */
+  virtual MLChargeAmount ChargeForward()
+  {
+    // TODO(VH): make a pure virtual call;
+    std::cout << "Error: call to unexisting ChargeForward() implementation! returned 0."
+              << std::endl;
+    return 0;
+  }
+
+  /**
+   * @brief ChargeBackward
+   * @return estimated charge amount, necessary for performing a backward pass on data of given
+   * shapes.
+   */
+  virtual MLChargeAmount ChargeBackward()
+  {
+    // TODO(VH): make a pure virtual call;
+    std::cout << "Error: call to unexisting ChargeBackward() implementation! returned 0."
+              << std::endl;
+    return 0;
+  }
+
+  /**
+   * @brief TotalElementsIn calculated a sum of total elements in all given tensors
+   * @param shapes - vector of tensor shapes
+   */
+  static fetch::math::SizeType TotalElementsIn(std::vector<fetch::math::SizeVector> const &shapes)
+  {
+    if (shapes.empty())
+    {
+      return 0;
+    }
+    fetch::math::SizeType total_elements = 1;
+    for (auto const &shape : shapes)
+    {
+      for (auto const &dimension : shape)
+      {
+        if (dimension == 0)
+        {
+          std::runtime_error("A dimension of size 0 found in tensor shape!");
+        }
+        total_elements *= dimension;
```

Comment:
is this correct? so if i have a custom op with three bias matrices that are just added to the incoming signal for example, then this will compute the total elements to be:
1 * n_elements in bias_1, and then that figure multiplied by dimension of bias_2, and then that figure multiplied by dimension of bias_3. But the number of elements is actually just dimension of bias_1 + d of bias_2 + d of bias_3 in that case, right?

I guess, i don't understand why ops.hpp's implementation isn't pure virtual anyway?

Reply:
yes, that is a mistake! sure it should multiply dimensions only inside one of the shapes and then sum up across different shapes. The reason for a static method are: 1) it is used only in derivatives of Ops, 2) it is exactly the same always, because it just calculates how much element would be in a tensor of some given shape. 3) it does not use any internal state of a particular Op